kliff/trainer/kim_trainer.py

```diff
@@ -0,0 +1,181 @@
+        function. It seems like MPI might be only way to make it parallel as the
+        multiprocessing does not work with the KIM models. KIMPY models are not yet
+        pickelable. TODO: include MPI support.
+        """
+        # set the parameters
+        self.model.update_model_params(x)
+        # compute the loss
+        loss = 0.0
+        for configuration in self.train_dataset:
+            compute_energy = True if configuration.weight.energy_weight else False
+            compute_forces = True if configuration.weight.forces_weight else False
+            compute_stress = True if configuration.weight.stress_weight else False
+
+            prediction = self.model(
+                configuration,
+                compute_energy=compute_energy,
+                compute_forces=compute_forces,
+                compute_stress=compute_stress,
+            )
+
+            if configuration.weight.energy_weight:
+                loss += configuration.weight.energy_weight * self.loss_function(
+                    prediction["energy"], configuration.energy
+                )
+            if configuration.weight.forces_weight:
+                loss += configuration.weight.forces_weight * self.loss_function(
+                    prediction["forces"], configuration.forces
+                )
+            if configuration.weight.stress_weight:
+                loss += configuration.weight.stress_weight * self.loss_function(
+                    prediction["stress"], configuration.stress
+                )
+            loss *= configuration.weight.config_weight
+
+        return loss
+
+    def checkpoint(self, *args, **kwargs):
+        TrainerError("checkpoint not implemented.")
+
+    def train_step(self, *args, **kwargs):
+        TrainerError("train_step not implemented.")
+
+    def validation_step(self, *args, **kwargs):
+        TrainerError("validation_step not implemented.")
+
+    def get_optimizer(self, *args, **kwargs):
+        TrainerError("get_optimizer not implemented.")
+
+    def train(self, *args, **kwargs):
+        def _wrapper_func(x):
+            return self.loss(x)
+
+        x = self.model.get_opt_params()
+        options = self.optimizer_manifest.get("kwargs", {})
+        options["options"] = {
+            "maxiter": self.optimizer_manifest["epochs"],
+            "disp": self.current["verbose"],
+        }
+        result = self.optimizer(
+            _wrapper_func, x, method=self.optimizer_manifest["name"], **options
+        )
+
+        if result.success:
+            logger.info(f"Optimization successful: {result.message}")
+            self.model.update_model_params(result.x)
+        else:
+            logger.error(f"Optimization failed: {result.message}")
+
+    def _get_loss_fn(self):
+        if self.loss_manifest["function"].lower() == "mse":
+            return MSE_residuals
+
+    def save_kim_model(self):
+        if self.export_manifest["model_type"].lower() == "kim":
```

Comment:
To simplify it, I don't think we need to distinguish `kim` and `tar`, because it seems the only difference is that the latter is the tarred version of the former. This requires change of some other files as well. 

I think we can generally call this a `kim` model. We have the option to tar it or not here. But when reading the kim model (in another place), we can use a utility function to check what is provided and then proceed accordingly. To be more specific, I can think of three modes: 
1. If a simple model name like `SW_xxxxxx` is provided, this will be regarded as the user wanting to use a model in a kim collection (e.g. user); 
2. If a path is provided like, `./SW_xxxxxx`, then the local model at the corresponding path will be installed and used;
3. If a tar path is provided like `./SW_xxxxxx.tgz`, we untar and then install. 

For the three models, we can internally use a utility function check whether it is a name, a path to a directory, or a path to a tarball.